File: src/main/java/de/ckthomas/smart/iot/services/RestServiceClient.java
package de.ckthomas.smart.iot.services;

import okhttp3.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;

/**
 * TODO refactor this part to kotlin using the spring techniques
 *
 * This facade provides common logic for a simpler usage.
 * @author Christian Thomas
 */
public class RestServiceClient {

    private static final Logger logger = LoggerFactory.getLogger(RestServiceClient.class);

    public static final MediaType JSON = MediaType.get("application/json; charset=utf-8");

    /**
     * use only one instance of the http-client.
     */
    private final OkHttpClient httpClient = new OkHttpClient();

    private final String basePath;
    private final String authKey;
    private final String authValue;

    /**
     * Use the factory to get an instance of this.
     */
    RestServiceClient(String basePath, String authKey, String authValue) {
        this.basePath = basePath;
        this.authKey = authKey;
        this.authValue = authValue;
    }

    public Request.Builder createRequestBuilder(String url) {
        final String prefix = basePath != null ? basePath + "/" : "";
        final String finalUrl = prefix + url;
        logger.info("Using final url = {} for creating request builder", finalUrl);

        final Request.Builder builder = new Request.Builder()
                .url(finalUrl);
        if (authKey != null && authValue != null) {
            logger.info("Adding header with authKey = {}, authValue = {}", authKey, "***secret***");
            builder.addHeader(authKey, authValue);
        }
        return builder;
    }

    public Request createPostRequest(String url, String jsonBody) {
        return createRequestBuilder(url)
            .post(RequestBody.create(jsonBody, JSON))
            .build();
    }

    public Call createCall(Request request) {
        return httpClient.newCall(request);
    }

    public Response execute(Request request) throws IOException {
        return createCall(request).execute();
    }

    public Response execute(String url, String jsonBody) throws IOException {
        return execute(createPostRequest(url, jsonBody));
    }

}

File: src/main/kotlin/de/ckthomas/smart/iot/services/RestServiceClient.kt
package de.ckthomas.smart.iot.services

import okhttp3.Request.Builder

import okhttp3.*
import org.slf4j.LoggerFactory

import kotlin.Throws
import java.io.IOException

/**
 * This facade provides common logic for a simpler usage.
 * @author Christian Thomas
 */
class RestServiceClient
/**
 * Use the factory to get an instance of this.
 */ internal constructor(private val basePath: String?, private val authKey: String?, private val authValue: String?) {
    /**
     * use only one instance of the http-client.
     */
    private val httpClient = OkHttpClient()
    fun createRequestBuilder(url: String): Builder {
        val prefix = if (basePath != null) "$basePath/" else ""
        val finalUrl = prefix + url
        LOGGER.info("Using final url = {} for creating request builder", finalUrl)
        val builder: Builder = Builder()
            .url(finalUrl)
        if (authKey != null && authValue != null) {
            LOGGER.info("Adding header with authKey = {}, authValue = {}", authKey, "***secret***")
            builder.addHeader(authKey, authValue)
        }
        return builder
    }

    fun createPostRequest(url: String, jsonBody: String?): Request {
        return createRequestBuilder(url)
            .post(RequestBody.create(jsonBody, JSON))
            .build()
    }

    fun createCall(request: Request?): Call {
        return httpClient.newCall(request!!)
    }

    @Throws(IOException::class)
    fun execute(request: Request?): Response {
        return createCall(request).execute()
    }

    @Throws(IOException::class)
    fun execute(url: String, jsonBody: String?): Response {
        return execute(createPostRequest(url, jsonBody))
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(RestServiceClient::class.java)
        val JSON: MediaType = get.get("application/json; charset=utf-8")
    }
}